keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1328 @@
+volume source that gives workloads easy filesystem-based access to trust anchor
+sets that they might need.
+
+## Motivation
+
+In the absence of a standard mechanism for trust distribution, a few different
+approaches have emerged:
+
+* The trust anchors necessary for connecting to kube-apiserver are distributed
+  using a ConfigMap that is automatically injected into every namespace.
+
+* Kubernetes webhooks and aggregated API servers are configured with a static
+  trust anchor set embedded in their config (which complicates rotation).
+
+* * At least one signer implementation uses a [cluster-scoped
+  CRD](https://cloud.google.com/traffic-director/docs/security-proxyless-setup#:~:text=Save%20the%20following%20TrustConfig%20YAML%20configuration%20to%20tell%20your%20cluster%20how%20to%20trust%20the%20issued%20certificates%3A)
+  to distribute trust anchors to workloads.
+
+Each of these approaches has downsides:
+
+* Creating a ConfigMap in every namespace requires granting over-broad
+  permissions to the responsible controller (create ConfigMaps in **all**
+  namespaces).
+
+* Static trust anchor sets embedded directly in configuration complicate the job
+  of rotating the backing CA's roots of trust.
+
+* CRDs work well, but are inappropriate for use by core APIs.
+
+Each of these approaches would be simplified by an in-tree mechanism for
+distributing trust anchor sets.
+
+### Goals
+
+* Provide a Kubernetes data type that holds X.509 trust anchors for use wherever
+  a set of X.509 trust anchors is needed.
+
+* Make it easy for certificates/v1 signer implementations to publish their
+  relevant trust anchors for use within the cluster.
+
+* Make it easy for workloads that need access to the trust anchors that back a
+  particular certificates/v1 signer to access them within the container
+  filesystem.
+
+* Allow cluster operators to configure webhook and aggregated API server
+  backends using that use trust anchor objects instead of inline PEM data.
+
+### Non-Goals
+
+* Distribute a default set of "system" trust anchors.  (This is a natural
+  extension, though).
+
+* Handle trust anchors expressed in other forms than PEM-wrapped, DER-formatted
+  X.509.
+
+## Proposal
+
+This proposal is centered around a new cluster-scoped ClusterTrustBundle
+resource, initially in the certificates/v1alpha1 API group.  The
+ClusterTrustBundle object can be thought of as a specialized configmap tailored
+to the X.509 trust anchor use case.  Introducing a dedicated type allows us to
+attach different RBAC policies to ClusterTrustBundle objects, which will
+typically be wide-open for reading, but locked-down for writing.
+
+Each ClusterTrustBundle object is a container for multiple named trust anchor
+sets.  Most use cases will simply use a single trust anchor set named `default`,
+but mutual TLS use cases involving multiple federating meshes will require the
+ability to express multiple, non-commingled trust anchor sets.
```

Comment:
As I understand it, federation between multiple corporate entities might involve multiple "Trust Domains" and multiple "Trust Anchors" for each domain, i.e. you have partners FooCorp and BarConglomorate and each has kubernetes clusters providing workload identity in "us-east-1" and "us-west-2" - The same GRPC Server is expected to trust both "FooCorp:us-east-1:workload" and "BarConglomorate:us-east-1:workload", and still differentiate the two.

It's unclear to me whether a complex mapping of multiple TrustAnchor certificates to their respective federation domains belongs in a single TrustAnchorSet, or if it would be better expressed as multiple AnchorSets with projection handling the virtual layout. At present, I don't think this is a design goal for v1.

Reply:
I'm going to drop the federation piece --- convert ClusterTrustBundle to just be a single set of trust anchors.  I think federation can be adequately handled with a "note" or "version" field in ClusterTrustBundle .spec

Reply:
Done.